### PR TITLE
fix function registerMember (Subs-Members.php)

### PR DESCRIPTION
**Pull request by javinotfound**
registerMember doesn't check if username (member_name) exists before allow to register it.

### CODE REVIEW

**Comment by albertlast** [comment]
i don't understand why you take the mail account check out,
why you create a second query.
could be done by one query by extend the where or use union.

**Comment by javinotfound** [comment]
The problem is that as it was the registerMember function a query is made to see if the email is already being used by another member, but it is not checked at any time if the username is already being used by another user.

Although I know that it could be done through a single query, when reporting which is the exact error (if it is the username that already exists, or on the contrary is the email address) I chose to make 2 queries.

In the current code it seems that the query checks that the user name entered is unique, but this does not happen because in any case 'OR email_address = {string:username}' should be 'OR member_name= {string:username}'.

**Comment by albertlast** [comment]
The problem is clear,
but the solution is unclear,
i would keep the mail check and i would keep this at one query not two.

**Comment by javinotfound** [comment]
The mail check is present in my fix (line 520, email is checked)

I deleted line 521 because it was a mistake (it should be 'OR member_name = {string:username}' instead of 'OR email_address = {string:username}'

I decided make two query instead of one because if you keep that in only one query, when a row is found, it could be username or email. You don't know if the problem is within the email or with the username. You will know that some of them is being used by another member, but no if it is the username or the email.

**Comment by albertlast** [comment]
sure you can know this,
by using case in the colum or use union or check this in php.

**Comment by javinotfound** [comment]
I edited the fix to execute only one query. Now it's ok?

File: Sources/Subs-Members.php [hunk 1]
```diff
@@ -513,22 +513,24 @@ function registerMember(&$regOptions, $return_errors = false)
...
 
-	// Check if the email address is in use.
+	// Check if the email address or the username is in use.
 	$request = $smcFunc['db_query']('', '
-		SELECT id_member
+		SELECT member_name
 		FROM {db_prefix}members
-		WHERE email_address = {string:email_address}
-			OR email_address = {string:username}
+		WHERE email_address = {string:email_address} OR member_name = {string:username}
```
Comment (albertlast) [inline Sources/Subs-Members.php]: get this back the old stylew and change only the member_name column

File: Sources/Subs-Members.php [hunk 2]
```diff
@@ -513,22 +513,24 @@ function registerMember(&$regOptions, $return_errors = false)
...
 		LIMIT 1',
-		array(
-			'email_address' => $regOptions['email'],
-			'username' => $regOptions['username'],
-		)
+		array('email_address' => $regOptions['email'], 'username' => $regOptions['username'])
```
Comment (albertlast) [inline Sources/Subs-Members.php]: get this back to the old style

File: Sources/Subs-Members.php [hunk 3]
```diff
@@ -513,22 +513,24 @@ function registerMember(&$regOptions, $return_errors = false)
...
 	);
+	
 	// @todo Separate the sprintf?
-	if ($smcFunc['db_num_rows']($request) != 0)
-		$reg_errors[] = array('lang', 'email_in_use', false, array($smcFunc['htmlspecialchars']($regOptions['email'])));
-
+	if ($smcFunc['db_num_rows']($request) != 0) {
+		list ($membername_db) = $smcFunc['db_fetch_row']($request);
+		if ($membername_db===$regOptions['username']) {
```
Comment (albertlast) [inline Sources/Subs-Members.php]: add a space on both side of ``===``

**Review by albertlast — CHANGES_REQUESTED** [review]

File: Sources/Subs-Members.php (hunk 1 above)
Comment (javinotfound) [reply Sources/Subs-Members.php]: Done

File: Sources/Subs-Members.php (hunk 2 above)
Comment (javinotfound) [reply Sources/Subs-Members.php]: Done

File: Sources/Subs-Members.php (hunk 3 above)
Comment (javinotfound) [reply Sources/Subs-Members.php]: Done

**Comment by javinotfound** [comment]
All changes suggested done.

File: Sources/Subs-Members.php [hunk 4]
```diff
@@ -513,22 +513,28 @@ function registerMember(&$regOptions, $return_errors = false)
...
 
-	// Check if the email address is in use.
+	// Check if the email address or the username is in use.
 	$request = $smcFunc['db_query']('', '
-		SELECT id_member
+		SELECT member_name
 		FROM {db_prefix}members
-		WHERE email_address = {string:email_address}
-			OR email_address = {string:username}
+		WHERE email_address = {string:email_address}
```
Comment (SychO9) [inline Sources/Subs-Members.php]:
```suggestion
		WHERE email_address = {string:email_address}
```

File: Sources/Subs-Members.php [hunk 5]
```diff
@@ -513,22 +513,28 @@ function registerMember(&$regOptions, $return_errors = false)
...
+		SELECT member_name
 		FROM {db_prefix}members
-		WHERE email_address = {string:email_address}
-			OR email_address = {string:username}
+		WHERE email_address = {string:email_address} 
+				OR member_name = {string:username}
 		LIMIT 1',
 		array(
-			'email_address' => $regOptions['email'],
+			'email_address' => $regOptions['email'],
```
Comment (SychO9) [inline Sources/Subs-Members.php]:
```suggestion
			'email_address' => $regOptions['email'],
```

File: Sources/Subs-Members.php [hunk 6]
```diff
@@ -513,22 +513,28 @@ function registerMember(&$regOptions, $return_errors = false)
...
+			'email_address' => $regOptions['email'], 
 			'username' => $regOptions['username'],
 		)
 	);
+	
 	// @todo Separate the sprintf?
-	if ($smcFunc['db_num_rows']($request) != 0)
-		$reg_errors[] = array('lang', 'email_in_use', false, array($smcFunc['htmlspecialchars']($regOptions['email'])));
-
+	if ($smcFunc['db_num_rows']($request) != 0) {
```
Comment (SychO9) [inline Sources/Subs-Members.php]:
```suggestion
	if ($smcFunc['db_num_rows']($request) != 0)
	{
```

File: Sources/Subs-Members.php [hunk 7]
```diff
@@ -513,22 +513,28 @@ function registerMember(&$regOptions, $return_errors = false)
...
+		list ($membername_db) = $smcFunc['db_fetch_row']($request);
+		if ($membername_db === $regOptions['username']) {
```
Comment (SychO9) [inline Sources/Subs-Members.php]:
```suggestion
		if ($membername_db === $regOptions['username'])
		{
```

File: Sources/Subs-Members.php [hunk 8]
```diff
@@ -513,22 +513,28 @@ function registerMember(&$regOptions, $return_errors = false)
...
+			$reg_errors[] = array('lang', 'already_a_user', false);
+		} else {
```
Comment (SychO9) [inline Sources/Subs-Members.php]:
```suggestion
		}
		else
		{
```

**Review by SychO9 — COMMENTED** [review]
Excuse the nitpicking, but just to keep the coding style the same as SMF's

Thank you for the fix, no one had noticed this before 😄

File: Sources/Subs-Members.php [hunk 9]
```diff
@@ -513,22 +513,32 @@ function registerMember(&$regOptions, $return_errors = false)
...
 
-	// Check if the email address is in use.
+	// Check if the email address or the username is in use.
 	$request = $smcFunc['db_query']('', '
-		SELECT id_member
+		SELECT member_name
 		FROM {db_prefix}members
 		WHERE email_address = {string:email_address}
-			OR email_address = {string:username}
+				OR member_name = {string:username}
```
Comment (albertlast) [inline Sources/Subs-Members.php]: The tab should be not here.

**Comment by javinotfound** [comment]
Sorry for not respecting SMF coding style. It's my first pull request on open source code.

All changes applied, waiting to travis.

File: Sources/Subs-Members.php [hunk 10]
```diff
@@ -1530,4 +1540,4 @@ function generateValidationCode()
 	return substr(preg_replace('/\W/', '', sha1(microtime() . $smcFunc['random_int']() . $dbRand . $modSettings['rand_seed'])), 0, 10);
 }
 
-?>
\ No newline at end of file
+?>
```
Comment (albertlast) [inline Sources/Subs-Members.php]: you use a different eol,
change it back to the old one.

**Review by albertlast — CHANGES_REQUESTED** [review]
to meet the code style stuff is one the first time always a strugle.

**Comment by SychO9** [comment]
It's allright, I remember my first pull request, it was a mess xD

**Comment by javinotfound** [comment]
I think that all works fine, respecting SMF coding style and tabulations.

**Review by albertlast — APPROVED** [review]

**Comment by albertlast** [comment]
I got notice that registerMember call validateUsername
https://github.com/SimpleMachines/SMF2.1/blob/e0848c82fbbc912b51789a2fefded6b6830f086f/Sources/Subs-Members.php#L476

 and in this the function isReservedName is called:
https://github.com/SimpleMachines/SMF2.1/blob/e0848c82fbbc912b51789a2fefded6b6830f086f/Sources/Subs-Auth.php#L681

so the username check should already work.
Without this pr.

**Comment by javinotfound** [comment]
I thought same as you, but for any reason I think that this function (isReservedName) only checks if username is reserved (for example if contains 'admin', 'root'... because this words are reserved by default), but not if username is unique and exists. You can take 2.1 RC2 release and with SSI.php, try to register two users with same username and different email, and you will see that registrationMember will work, creating two users with same username in the forum. Perhaphs we need to change error message  un isReservedName.

**Comment by albertlast** [comment]
For me the line look similiar what you do:
https://github.com/SimpleMachines/SMF2.1/blob/e0848c82fbbc912b51789a2fefded6b6830f086f/Sources/Subs-Members.php#L911-L930

So maybe this function got an issue which should get be fixed?

**Comment by MissAllSunday** [comment]
This seems very similar to what SMF already does, will close for now but please feel free to create another issue with an example of how to reproduce the behavior this PR solves.

**Pull request closed without merging**